pipeline/common/downloads.py

```diff
@@ -85,47 +78,233 @@ def get_download_size(url: str) -> int:
-        if mocked_stream:
+        mocked_request = attempt_mocked_request(self.url)
+        if mocked_request:
             # We are in a test.
             logger.info(f"Using a mocked download: {self.url}")
-            self.response = mocked_stream
+            self.byte_chunk_stream = mocked_request
+            self.decoding_stream = self.decode(self.byte_chunk_stream)
         else:
-            self.response = requests.get(self.url, stream=True)
-            self.response.raise_for_status()
+            self.byte_chunk_stream = DownloadChunkStreamer(self.url).__enter__()
+            self.decoding_stream = self.decode(self.byte_chunk_stream)
 
-        self.decoding_stream = self.decode(self.response.raw)
         self.line_stream = io.TextIOWrapper(self.decoding_stream, encoding="utf-8")
+
         return self.line_stream
 
     def __exit__(self, _exc_type, _exc_val, _exc_tb):
         self.line_stream.close()
         self.decoding_stream.close()
-        self.response.close()
+        self.byte_chunk_stream.close()
 
-    def decode(self):
-        raise NotImplementedError
+    def decode(self, byte_stream):
+        raise NotImplementedError("Decoding is not implemented")
 
 
 class RemoteGzipLineStreamer(RemoteDecodingLineStreamer):
-    """Stream lines directly from a remote gzip file."""
+    """
+    Stream lines directly from a remote gzip file. The line includes the newlines separator.
+
+    Usage:
 
-    def decode(self, response_stream):
-        return gzip.GzipFile(fileobj=response_stream)
+        with RemoteGzipLineStreamer(url) as lines:
+            for line in lines:
+                print(line)
+    """
+
+    def decode(self, byte_stream):
+        return gzip.GzipFile(fileobj=byte_stream)
 
 
 class RemoteZstdLineStreamer(RemoteDecodingLineStreamer):
-    """Stream lines directly from a remote zstd file."""
+    """
+    Stream lines directly from a remote zstd file. The line includes the newlines separator.
+
+    Usage:
+
+        with RemoteGzipLineStreamer(url) as lines:
+            for line in lines:
+                print(line)
+    """
+
+    def decode(self, byte_stream):
+        return zstandard.ZstdDecompressor().stream_reader(byte_stream)
+
+
+class DownloadChunkStreamer(io.IOBase):
+    """
+    Streams a download as chunks, and retries several times if there are any failures. This
+    clas implements io.IOBase so it can be used as a file reader.
+
+    Iterator over chunks directly:
+
+        with DownloadChunkStreamer(url) as chunk_streamer:
+            for chunk in chunk_streamer.download_chunks():
+                f.write(chunk)
 
-    def decode(self, response_stream):
-        return zstandard.ZstdDecompressor().stream_reader(response_stream)
+    Or pass it as a file handle:
+
+        with DownloadChunkStreamer(url) as f:
+             gzip.GzipFile(fileobj=f)
+    """
+
+    def __init__(self, url: str, total_retries=20, timeout_sec=10.0, wait_before_retry_sec=1.0):
```

Comment:
IMO it's better to reduce the number of retries to say 3 but increase the retry intervals to say 60 sec. Libs like tenacity also allow retry policies to exponentially increase the interval each time but it's optional. With 20 retries with a 1-second interval, you risk exceeding the request rate set by the external resource and being blocked.

Reply:
Thanks, I adjusted it.